frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -389,39 +417,68 @@ pub mod pallet {
                 Error::<T>::NotOwnedContract,
             );
 
-            // We need to unstake all funds that are currently staked
             let current_era = Self::current_era();
-            let staking_info = Self::staking_info(&contract_id, current_era);
-            for (staker, amount) in staking_info.stakers.iter() {
-                let mut ledger = Self::ledger(staker);
-                ledger.locked = ledger.locked.saturating_sub(*amount);
-                Self::update_ledger(staker, ledger);
-            }
+            dapp_info.state = DAppState::Unregistered(current_era);
+            RegisteredDapps::<T>::insert(&contract_id, dapp_info);
 
-            // Need to update total amount staked
-            let staking_total = staking_info.total;
-            EraRewardsAndStakes::<T>::mutate(
-                &current_era,
-                // XXX: RewardsAndStakes should be set by `on_initialize` for each era
-                |value| {
-                    if let Some(x) = value {
-                        x.staked = x.staked.saturating_sub(staking_total)
-                    }
-                },
-            );
-
-            // Nett to update staking data for next era
-            let empty_staking_info = EraStakingPoints::<T::AccountId, BalanceOf<T>>::default();
-            ContractEraStake::<T>::insert(contract_id.clone(), current_era, empty_staking_info);
-
-            // Developer account released but contract can not be released more.
+            // Developer account released.
             T::Currency::unreserve(&developer, T::RegisterDeposit::get());
+
+            // TODO: Imo this should be removed
             RegisteredDevelopers::<T>::remove(&developer);
 
             Self::deposit_event(Event::<T>::ContractRemoved(developer, contract_id));
 
-            let number_of_stakers = staking_info.stakers.len();
-            Ok(Some(T::WeightInfo::unregister(number_of_stakers as u32)).into())
+            Ok(().into())
+        }
+
+        /// Withdraw locked funds from a contract that was unregistered.
+        /// Funds don't need to undergo the unbonding period - they are returned immediately.
+        #[pallet::weight(T::WeightInfo::withdraw_from_unregistered())]
+        pub fn withdraw_from_unregistered(
```

Comment:
@bobo-k2 how do you plan to implement this on frontend?

Reply:
@Dinonard could you explain please why it should be dedicated method? It looks like a case for general unbond method.

Reply:
It's actually quite different - the checks differ.

Here we require that all pending rewards have been claimed. Also, there's no unbonding period here.

Sure, I could put it all into `unbond_and_unstake` but it creates a more complicated method with many **if-s**.
This, in my opinion, is much better. Code is easier to test and there is less room for error.

Reply:
Clean interface is also important. Not sure why withdraw from unregistered dapps should be different from general withdraw.

Reply:
I agree with you but I don't want to make code more complicated - by doing this split, there is no branching, most if/else clauses are avoided.

And the behavior is inherently different.

If we use `unbond_and_unstake` we can `withdaraw_unbonded` after unbonding period has passed.
However, `withdraw_from_unregistered` acts as `unbond_unstake_and_withdraw` all at once.
So the behavior is quite different.

There is no unbonding period here - funds are returned directly to the user.
Also, you cannot call this if there are pending rewards - this is one of the simplifications I made to avoid complexity in the `StakerInfo`.

Should we maybe rename it?